src/unix/ftrace.c

```diff
@@ -1475,3 +1768,176 @@ ftrace_deinit(void)
 {
     deallocate_http_listener(ftrace_heap, ftrace_hl);
 }
+
+int
+ftrace_thread_init(thread t)
+{
+    t->graph_stack = allocate(ftrace_heap,
+        sizeof(struct ftrace_graph_entry) * FTRACE_RETFUNC_DEPTH
+    );
+    if (t->graph_stack == INVALID_ADDRESS) {
+        msg_err("failed to allocare ftrace return stack array\n");
+        return -ENOMEM;
+    }
+
+    t->graph_idx = 0;
+    return 0;
+}
+
+void
+ftrace_thread_deinit(thread t)
+{
+    deallocate(ftrace_heap, t->graph_stack,
+        sizeof(struct ftrace_graph_entry) * FTRACE_RETFUNC_DEPTH
+    );
+
+    t->graph_idx = 0;
+    t->graph_stack = 0;
+}
+
+/*
+ * Reset call stack for newly running thread and record the switch event
+ */
+__attribute__((no_instrument_function))
+void
+ftrace_thread_switch(thread out, thread in)
+{
+    in->graph_idx = 0;
+
+    if (!rbuf_enabled(&global_rbuf))
+        return;
+
+    /* generate a ctx switch event if we're running function graph */
+    rbuf_disable(&global_rbuf);
+    if (runtime_strcmp(current_tracer->name, "function_graph") == 0)
+        function_graph_trace_switch(out, in);
+    rbuf_enable(&global_rbuf);
+}
+
+/* defined in src/x86_64/ftrace.s */
+extern void return_to_handler(void);
+
+/*
+ * Assume init_unix calls mcache_alloc
+ * self  : IP in mcache_alloc() where mcount is invoked
+ * parent: stack address where return address sits (from mcache_alloc
+ *      back into init_unix; i.e., *parent is an address in init_unix)
+ * frame_pointer: frame_pointer during call to mcache_alloc()
+ *      - i.e., *(frame_pointer+8) is the return address to get back
+ *       from init_unix() to its own parent (in this case, startup())
+ *
+ * What this function does is moves *parent so that mcache_alloc() returns
+ * to a stub of ours that can record the time, before jumping back into
+ * init_unix()
```

Comment:
Could there be an issue here with tail call elimination? Or will the compiler rule them out when -pg is specified?

Reply:
I doubt gcc would mcount() a function that it also tail optimizes -- but I will double check